pkg/storage/etcd3/util.go

```diff
@@ -0,0 +1,57 @@
+/*
+Copyright 2016 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package etcd3
+
+import (
+	"time"
+
+	"github.com/coreos/etcd/clientv3"
+	"github.com/golang/glog"
+	"golang.org/x/net/context"
+)
+
+// compactor periodically compacts historical versions of keys in etcd.
+// After compaction, old versions of keys set before given interval will be gone.
+// Any API call for the old versions of keys will return error.
```

Comment:
at some point, we need to compact the history. we should not allow the history to grow unboundedly. Right now, no consumer should even expect to consume historical data before 1000 or say 10000 versions due to etcd2's implementation. Also the storage API does not expose this details. We can pace the compaction in storage itself as long as the API stays the same. Once we do want to expose historical versions of keys, we might change the default behavior.